data/lua/cave_map_generator.lua

```diff
@@ -0,0 +1,205 @@
+		y = random(tonumber(y_min), tonumber(y_max))
+	end
+	local locs_set = {}
+	build_chamber(x, y, locs_set, chambers.size or 3, chambers.size or 0)
+	local items = {}
+	std_print("before item")
+	for item in helper.child_range(chamber, "item") do
+		std_print("added item")
+		table.insert(items, item)
+	end
+	table.insert(chambers, {
+		center_x = x,
+		center_y = y,
+		side_num = chamber.side,
+		locs_set = locs_set,
+		id = id,
+		items = items,
+		--items = helper.get_child(v, "items")
+	})
+	chambers_by_id[id] = chambers[#chambers]
+	for passage in helper.child_range(chamber, "passage") do
+		local dst = chambers_by_id[passage.destination]
+		if dst ~= nil then
+			table.insert(passages, {
+				start_x = x,
+				start_y = y,
+				dest_x = dst.center_x,
+				dest_y = dst.center_y,
+				data = passage,
+			})
+		end
+	end
+end
+
+for i,v in ipairs(chambers) do
+	std_print("processing chamber " .. (v.id or ""))
+	local locs_list = {}
+	for k2,v2 in pairs(v.locs_set) do
+		set_tile_index(k2, params.terrain_clear)
+		if on_inner_board (v2[1], v2[2]) then
+			table.insert(locs_list, k2)
+		end
+	end
+	for i1, item in ipairs(v.items or {}) do
+		local index = random(#locs_list)
+		local loc = locs_list[index]
+		table.remove(locs_list, index)
+		std_print("placed special location " .. (item.id or "") .. " at " .. loc)
+		local x, y = index_to_loc(loc)
+		
+		starting_positions[loc] = item.id
+		
+		if item.place_castle then
+			set_tile_index(loc, params.terrain_keep)
+			for x2, y2 in adjacent_tiles(x, y) do
+				set_tile(x2, y2, params.terrain_castle)
+			end
+		end
+	end
+end
+
+for i,v in ipairs(passages) do
+	if random(100) <= (v.data.chance or 100) then
+		local windiness = v.data.windiness or 0
+		local laziness = math.max(v.data.laziness or 1, 1)
+		local width = math.max(v.data.width or 1, 1)
+		local jagged = v.data.jagged or 0
+		local calc = function(x, y, current_cost)
+			local res = 1.0
+			if get_tile(x, y) == params.terrain_wall then
+				res = laziness
+			end
+			if windiness > 1 then
+				res = res * random(windiness)
+			end
+			return res
+		end
+		local path, cost = wesnoth.find_path(v.start_x, v.start_y, v.dest_x, v.dest_y, calc, params.map_width, params.map_height)
+		for i, loc in ipairs(path) do
+			local locs_set = {}
+			build_chamber(loc[1], loc[2], locs_set, width, jagged)
+			for k,v in pairs(locs_set) do
+				set_tile_index(k, params.terrain_clear)
+			end
+		end
+	end
+	
+end
+
+local stringbuilder = {}
+
+for y = 0, params.map_height - 1 do
+	local stringbuilder_row = {}
+	for x = 0, params.map_width - 1 do
+		local index = x + 1 + y * params.map_width
+		if starting_positions[index] ~= nil then
+			table.insert(stringbuilder_row, tostring(starting_positions[index]) .. " " .. map[index])
+			else
+			table.insert(stringbuilder_row, map[index])
+		end
+	end
+	table.insert(stringbuilder, table.concat(stringbuilder_row, ","))
+end
+std_print( table.concat(stringbuilder, "\n"))
+return table.concat(stringbuilder, "\n")
```

Comment:
That was an observation of the return value, actually. It would make more sense to me if you actually return an array of "tiles". But it's not a deal-breaker or anything.